django/utils/functional.py

```diff
@@ -18,13 +20,54 @@ class cached_property:
     Decorator that converts a method with a single self argument into a
     property cached on the instance.
 
-    Optional ``name`` argument allows you to make cached properties of other
-    methods. (e.g.  url = cached_property(get_absolute_url, name='url') )
+    A cached property can be made out of an existing method:
+    (e.g. ``url = cached_property(get_absolute_url)``).
+    On Python < 3.6, the optional ``name`` argument must be provided, e.g.
+    ``url = cached_property(get_absolute_url, name='url')``.
     """
+    name = None
+
+    @staticmethod
+    def func(instance):
+        raise TypeError(
+            'Cannot use cached_property instance without calling '
+            '__set_name__() on it.'
+        )
+
+    @staticmethod
+    def _is_mangled(name):
+        return name.startswith('__') and not name.endswith('__')
+
     def __init__(self, func, name=None):
-        self.func = func
+        if PY36:
+            self.real_func = func
+        else:
+            func_name = func.__name__
+            name = name or func_name
+            if self._is_mangled(func_name):
+                raise ValueError(
+                    'cached_property does not work with mangled methods on '
+                    'Python < 3.6 without the appropriate `name` argument. See '
+                    'https://docs.djangoproject.com/en/%s/ref/utils/'
+                    '#cached-property-mangled-name' % get_docs_version(),
+                )
+            if not (isinstance(name, str) and name.isidentifier()):
+                raise ValueError(
+                    "%r can't be used as name of cached_property" % name,
+                )
+            self.name = name
+            self.func = func
         self.__doc__ = getattr(func, '__doc__')
-        self.name = name or func.__name__
+
+    def __set_name__(self, owner, name):
+        if self.name is None:
+            self.name = name
+            self.func = self.real_func
+        elif name != self.name:
+            raise TypeError(
+                "Cannot assign the same cached_property to two different names "
+                "(%r and %r)." % (self.name, name)
+            )
```

Comment:
FWIW the `TypeError` raised here would be backward incompatible. Such a pattern is even used in Django core.

https://github.com/django/django/blob/0ce2ad9ca4623cfd6dc2515430c0ae8a1717a607/django/db/backends/postgresql/features.py#L56-L74

Reply:
Yea, as Sergey said above, there's a bug where such attributes aren't cached.

Reply:
Ahh right just read through the whole discussion.

I wonder if this could be an opportunity to solve the issue.

e.g.

```python
def __set_name__(self, owner, name):
    if self.name is None:
        self.name = name
        self.func = self.real_func
    elif name != self.name:
        # Re-assignment of a cached property to a class
        # creates a new instance.
        instance = type(self)(self.real_func, name)
        setattr(owner, name, instance)
```

That would make 

```python
@cached_property
def is_postgresql_9_5(self): 
    return self.connection.pg_version >= 90500

has_select_for_update_skip_locked = is_postgresql_9_5
```

The equivalent of

```python
@cached_property
def is_postgresql_9_5(self): 
    return self.connection.pg_version >= 90500

has_select_for_update_skip_locked = cached_property(
    is_postgresql_9_5.real_func, 'has_select_for_update_skip_locked'
)
```

And address the bug preventing the `__dict__` level caching discovered by @sergei-maertens.

Reply:
> I wonder if this could be an opportunity to solve the issue.

I considered it, but:

1. it doesn't fix caching on Python 3.5, raising `TypeError` on Python 3.6 could useful when code is tested against both versions (for example we wouldn't know that we use such pattern that works incorrectly on Python 3.5)
1. I'm not sure if we want to make it work this way, we might want this instead:

```python
@cached_property
def is_postgresql_9_5(self): 
    return self.connection.pg_version >= 90500

has_select_for_update_skip_locked = property(
    attrgetter('is_postgresql_9_5')
)
```

This way touching both `is_postgresql_9_5` and `has_select_for_update_skip_locked` will cause `is_postgresql_9_5.real_func` be called just once.
We might add `cached_property.alias()` method for functionality.